openwillis/measures/video/facial_emotion.py

```diff
@@ -5,19 +5,159 @@
+
+    arguments:
+    detector: pyfeat detector object
+    frame: frame to detect faces in
+    bb_dict: dictionary with bounding box coordinates
+    threshold: threshold for face detection
+
+    returns:
+    faces: list of faces detected in frame
+    '''
+    if len(bb_dict.keys()):
+        faces = bb_dict_to_bb_list(bb_dict)
+    else:
+        faces = detector.detect_faces(
+                frame,
+                threshold=.95,
+            )
+    return faces
+
+def mouth_openness(
+    landmarks,
+    upper_lip_lmks=[61, 62, 63],
+    lower_lip_lmks=[65, 66, 67]
+):
+    """
+    ------------------------------------------------------------------------------------------------------
+    This function calculates the average distance between the upper and lower lip landmarks to measure the
+    openness of the mouth.
+
+    Parameters:
+    ..........
+    landmarks: numpy array
+        An array containing the facial landmarks detected by the facial landmark detector.
+    upper_lip_lmks: list
+        A list containing the indices of the landmarks that make up the upper lip.
+    lower_lip_lmks: list
+        A list containing the indices of the landmarks that make up the lower lip.
+    
+    Returns:
+    ..........
+    lmk_dist: float
+        The average distance between the upper and lower lip landmarks.
+    ------------------------------------------------------------------------------------------------------
+    """
+
+    upper_lip = landmarks[upper_lip_lmks]
+    lower_lip = landmarks[lower_lip_lmks]
+
+    lmk_dist = []
+    for (upper_lip_x, upper_lip_y), (lower_lip_x, lower_lip_y) in zip(upper_lip, lower_lip):
+        lmk_dist.append(np.sqrt((upper_lip_x - lower_lip_x)**2 + (upper_lip_y - lower_lip_y)**2))
+    return np.mean(lmk_dist)
+
+
+def detect_emotions(detector, frame, emo_cols, bb_dict={},threshold=.95):
+
+    faces = get_faces(
+        detector,
+        frame,
+        bb_dict,
+        threshold=threshold
+    )
+
+    if len(faces[0])<1:
+        raise ValueError('No faces detected in frame')
+    
+    landmarks = detector.detect_landmarks(
+        frame,
+        detected_faces=faces
+    )
 
-def run_deepface(path, measures):
+    aus = detector.detect_aus(frame, landmarks)
+
+    emotions = detector.detect_emotions(
+        frame,
+        faces,
+        landmarks
+    )
+
+    emotions = emotions[0][0] * 100 # convert from 0-1 to 0-100
+    aus = aus[0][0]
+    landmarks = landmarks[0][0]
+
+    emos_aus = np.hstack([emotions,aus])
+    df_emo = pd.DataFrame([emos_aus],columns=emo_cols)
+    df_emo['mouth_openness'] = mouth_openness(landmarks)
+
+    return df_emo
+
+def run_pyfeat(path, skip_frames=5, bbox_list=[]):
     """
     ------------------------------------------------------------------------------------------------------
     This function takes an image path and measures config object as input, and uses the DeepFace package to
```

Comment:
Replace DeepFace with py-feat?

Reply:
Yep! This is just a q right? No suggestion?